File: vscode/src/graph.ts
// Microsoft Graph auth + requests, for the Entra ID (directory) PIM endpoints.
//
// The Azure CLI cannot help here. Its first-party app is not preauthorized for
// the Graph role-management scopes, so `az login --scope
// https://graph.microsoft.com/RoleManagement.Read.Directory ...` is rejected
// outright with AADSTS65002, and a plain CLI token is refused by Graph with
// PermissionScopeNotGranted.
//
// Two ways out, tried in that order:
//   1. VS Code's built-in Microsoft account provider — no code to type, and it
//      can be pointed at your own app registration through the VSCODE_CLIENT_ID
//      scope. Its default client may still be refused the PIM scopes.
//   2. Our own device-code flow against a public client that *is* preauthorized
//      for the delegated Graph surface, with the refresh token in SecretStorage.
//
// Azure *resource* roles need none of this — see pim.ts, which reaches ARM with
// a token from `az account get-access-token`.

import * as vscode from "vscode";

export const GRAPH = "https://graph.microsoft.com/v1.0";

const AUTHORITY = "https://login.microsoftonline.com";

/** Microsoft Graph PowerShell's public client id — preauthorized tenant-wide. */
export const DEFAULT_CLIENT_ID = "14d82eec-204b-4c2f-b7e8-296a70dab67e";

const SCOPES = [
  "https://graph.microsoft.com/RoleEligibilitySchedule.Read.Directory",
  "https://graph.microsoft.com/RoleAssignmentSchedule.ReadWrite.Directory",
  "https://graph.microsoft.com/RoleManagement.Read.Directory",
];

const SECRET_KEY = "azpim.graph.refreshToken";
const VSCODE_AUTH_REJECTED = "azpim.graph.vscodeAuthRejected";

export interface DeviceCode {
  userCode: string;
  verificationUri: string;
  expiresIn: number;
}

export class GraphError extends Error {
  constructor(
    message: string,
    readonly missingScope = false,
  ) {
    super(message);
  }
}

function config() {
  const c = vscode.workspace.getConfiguration("azpim");
  return {
    mode: c.get<"auto" | "vscode" | "device">("entraAuth", "auto"),
    clientId: c.get<string>("graphClientId")?.trim() || DEFAULT_CLIENT_ID,
    tenantId: c.get<string>("graphTenantId")?.trim() || "",
  };
}

/** Graph says this when the token lacks the role-management scopes. */
function looksLikeMissingScope(message: string): boolean {
  return /PermissionScopeNotGranted|Authorization_RequestDenied|insufficient privileges/i.test(message);
}

export class GraphAuth {
  /** Set once the extension knows which tenant `az` is pointed at. */
  cliTenantId: string | undefined;

  /** Replaced by the UI so the code can be shown in the sidebar. */
  onDeviceCode: (info: DeviceCode) => void = (info) => {
    void vscode.window.showWarningMessage(`Azure PIM: open ${info.verificationUri} and enter code ${info.userCode}`);
  };

  onAuthenticated: () => void = () => {};

  private accessToken: string | undefined;
  private expiresAt = 0;
  /** Which flow produced the current access token. */
  private source: "vscode" | "device" | undefined;
  private inFlight: Promise<string> | undefined;

  constructor(
    private readonly secrets: vscode.SecretStorage,
    private readonly memento: vscode.Memento,
  ) {}

  /** True when a token is already held or cached, i.e. no prompt is imminent. */
  async authenticated(): Promise<boolean> {
    if (this.accessToken) {
      return true;
    }
    if (await this.secrets.get(SECRET_KEY)) {
      return true;
    }
    if (config().mode !== "device" && !this.memento.get<boolean>(VSCODE_AUTH_REJECTED)) {
      const session = await vscode.authentication.getSession("microsoft", this.vscodeScopes(), { silent: true });
      return session !== undefined;
    }
    return false;
  }

  /**
   * A Graph access token. Concurrent callers share one acquisition — otherwise
   * the first refresh would start several device-code flows at once.
   */
  token(): Promise<string> {
    if (this.accessToken && this.expiresAt - 60_000 > Date.now()) {
      return Promise.resolve(this.accessToken);
    }
    if (!this.inFlight) {
      this.inFlight = this.acquire().finally(() => {
        this.inFlight = undefined;
      });
    }
    return this.inFlight;
  }

  /** Give up on the token in hand; the next call acquires a fresh one. */
  async invalidate(): Promise<void> {
    if (this.source === "vscode") {
      // The provider's token was refused the PIM scopes — stop asking it.
      await this.memento.update(VSCODE_AUTH_REJECTED, true);
    }
    this.accessToken = undefined;
    this.expiresAt = 0;
    this.source = undefined;
  }

  /** True when there is another flow left to try after invalidate(). */
  canRetryWithAnotherFlow(): boolean {
    return this.source === "vscode" && config().mode === "auto";
  }

  async logout(): Promise<void> {
    this.accessToken = undefined;
    this.expiresAt = 0;
    this.source = undefined;
    await this.secrets.delete(SECRET_KEY);
    await this.memento.update(VSCODE_AUTH_REJECTED, undefined);
  }

  // -------------------------------------------------------------------------

  private vscodeScopes(): string[] {
    const { clientId, tenantId } = config();
    const scopes = [...SCOPES];
    // The built-in provider understands these pseudo-scopes; they let it sign
    // in with your own registration instead of VS Code's own client.
    if (clientId !== DEFAULT_CLIENT_ID) {
      scopes.push(`VSCODE_CLIENT_ID:${clientId}`);
    }
    const tenant = tenantId || this.cliTenantId;
    if (tenant) {
      scopes.push(`VSCODE_TENANT:${tenant}`);
    }
    return scopes;
  }

  private async acquire(): Promise<string> {
    const { mode } = config();
    const rejected = this.memento.get<boolean>(VSCODE_AUTH_REJECTED) === true;

    if (mode !== "device" && !(mode === "auto" && rejected)) {
      try {
        const scopes = this.vscodeScopes();
        // A session may already exist for these scopes; only prompt if not.
        const session =
          (await vscode.authentication.getSession("microsoft", scopes, { silent: true })) ??
          (await vscode.authentication.getSession("microsoft", scopes, { createIfNone: true }));
        if (session) {
          this.accessToken = session.accessToken;
          // The provider refreshes on its own; re-ask often enough to notice.
          this.expiresAt = Date.now() + 10 * 60_000;
          this.source = "vscode";
          this.onAuthenticated();
          return session.accessToken;
        }
        if (mode === "vscode") {
          throw new GraphError("VS Code's Microsoft account provider returned no session");
        }
      } catch (e) {
        if (mode === "vscode") {
          throw e instanceof GraphError ? e : new GraphError(String((e as Error).message ?? e));
        }
        // The provider cannot get the PIM scopes here — remember that, so the
        // next refresh goes straight to the device-code flow.
        await this.memento.update(VSCODE_AUTH_REJECTED, true);
      }
    }

    const token = await this.deviceOrRefresh();
    this.source = "device";
    this.onAuthenticated();
    return token;
  }

  private async deviceOrRefresh(): Promise<string> {
    const refresh = await this.secrets.get(SECRET_KEY);
    if (refresh) {
      try {
        return await this.refreshFlow(refresh);
      } catch {
        // Expired or revoked: sign in again rather than fail.
        await this.secrets.delete(SECRET_KEY);
      }
    }
    return this.deviceFlow();
  }

  private async tenant(): Promise<string> {
    return config().tenantId || this.cliTenantId || "organizations";
  }

  private async post(path: string, form: Record<string, string>): Promise<any> {
    const res = await fetch(`${AUTHORITY}/${await this.tenant()}${path}`, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams(form).toString(),
    });
    const text = await res.text();
    try {
      return JSON.parse(text);
    } catch {
      throw new GraphError(`unexpected sign-in response (HTTP ${res.status}): ${text.slice(0, 300)}`);
    }
  }

  private store(data: any): string {
    this.accessToken = data.access_token;
    this.expiresAt = Date.now() + (Number(data.expires_in) || 3600) * 1000;
    if (data.refresh_token) {
      // SecretStorage keeps this in the OS keychain — it can mint Graph tokens.
      void this.secrets.store(SECRET_KEY, data.refresh_token);
    }
    return this.accessToken!;
  }

  private async refreshFlow(refreshToken: string): Promise<string> {
    const data = await this.post("/oauth2/v2.0/token", {
      grant_type: "refresh_token",
      client_id: config().clientId,
      refresh_token: refreshToken,
      scope: [...SCOPES, "offline_access"].join(" "),
    });
    if (!data.access_token) {
      throw new GraphError(data.error_description || data.error || "refresh failed");
    }
    return this.store(data);
  }

  private async deviceFlow(): Promise<string> {
    const clientId = config().clientId;
    const start = await this.post("/oauth2/v2.0/devicecode", {
      client_id: clientId,
      scope: [...SCOPES, "offline_access"].join(" "),
    });
    if (!start.device_code) {
      throw new GraphError(start.error_description || start.error || "could not start device-code sign-in");
    }
    this.onDeviceCode({
      userCode: start.user_code,
      verificationUri: start.verification_uri ?? "https://login.microsoft.com/device",
      expiresIn: Number(start.expires_in) || 900,
    });

    const deadline = Date.now() + (Number(start.expires_in) || 900) * 1000;
    let interval = (Number(start.interval) || 5) * 1000;
    for (;;) {
      await new Promise((r) => setTimeout(r, interval));
      const data = await this.post("/oauth2/v2.0/token", {
        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
        client_id: clientId,
        device_code: start.device_code,
      });
      if (data.access_token) {
        return this.store(data);
      }
      if (data.error === "authorization_pending" || data.error === "slow_down") {
        if (Date.now() > deadline) {
          throw new GraphError("device-code sign-in timed out");
        }
        if (data.error === "slow_down") {
          interval += 5000;
        }
        continue;
      }
      throw new GraphError(data.error_description || data.error || "device-code sign-in failed");
    }
  }
}

export class GraphClient {
  constructor(readonly auth: GraphAuth) {}

  /** One Graph call, retrying once if a token turns out to lack the scopes. */
  async request(method: string, url: string, body?: unknown, retried = false): Promise<any> {
    const token = await this.auth.token();
    const res = await fetch(url, {
      method: method.toUpperCase(),
      headers: {
        Authorization: `Bearer ${token}`,
        "Accept-Language": "en-US",
        ...(body === undefined ? {} : { "Content-Type": "application/json" }),
      },
      body: body === undefined ? undefined : JSON.stringify(body),
    });
    const text = await res.text();
    let data: any = {};
    if (text.trim()) {
      try {
        data = JSON.parse(text);
      } catch {
        throw new GraphError(`could not parse Graph response as JSON: ${text.slice(0, 300)}`);
      }
    }
    if (data.error || !res.ok) {
      const message = graphErrorMessage(data) ?? `Graph returned HTTP ${res.status}`;
      const missing = looksLikeMissingScope(message) || res.status === 403;
      if (missing && !retried && this.auth.canRetryWithAnotherFlow()) {
        await this.auth.invalidate();
        return this.request(method, url, body, true);
      }
      if (missing) {
        await this.auth.invalidate();
      }
      throw new GraphError(message, missing);
    }
    return data;
  }

  /** GET a URL, following `@odata.nextLink` until exhausted. */
  async getAll(url: string): Promise<any[]> {
    const out: any[] = [];
    let next: string | undefined = url;
    while (next) {
      const data = await this.request("get", next);
      out.push(...(data.value ?? []));
      next = data["@odata.nextLink"];
    }
    return out;
  }
}

function graphErrorMessage(data: any): string | undefined {
  const e = data?.error;
  if (e && typeof e === "object") {
    return e.message || e.code || JSON.stringify(e);
  }
  if (typeof e === "string") {
    return data.error_description || e;
  }
  return undefined;
}

export const GRAPH_SCOPE_HINT =
  "Entra ID roles need the Graph role-management scopes consented for the client in use. " +
  "A Global Admin must grant them — see the README. Then run “Azure PIM: Forget Cached Entra ID Sign-in” and retry.";

File: lua/azpim/graph.lua
-- Microsoft Graph auth + requests, for the Entra ID (directory) PIM endpoints.
--
-- The Azure CLI cannot help here. Its first-party app is not preauthorized for
-- the Graph role-management scopes, so `az login --scope
-- https://graph.microsoft.com/RoleManagement.Read.Directory ...` is rejected
-- outright with AADSTS65002, and a plain CLI token is refused by Graph with
-- PermissionScopeNotGranted. We therefore run our own device-code flow against
-- a public client that *is* preauthorized for the delegated Graph surface, and
-- cache the resulting refresh token.
--
-- Azure *resource* roles need none of this — see az.lua, which reaches ARM
-- through `az rest` with whatever `az login` already gave you.
local http = require("azpim.http")

local M = {}

local AUTHORITY = "https://login.microsoftonline.com"

--- Microsoft Graph PowerShell's public client id. Override via
--- `require("azpim").setup({ graph_client_id = ... })` if your tenant blocks it
--- and you have registered your own public client instead.
M.client_id = "14d82eec-204b-4c2f-b7e8-296a70dab67e"

local SCOPES = {
  "https://graph.microsoft.com/RoleEligibilitySchedule.Read.Directory",
  "https://graph.microsoft.com/RoleAssignmentSchedule.ReadWrite.Directory",
  "https://graph.microsoft.com/RoleManagement.Read.Directory",
  "offline_access",
}

--- Called when an interactive sign-in is needed, with
--- `{ user_code, verification_uri, expires_in }`. The UI replaces this.
function M.on_device_code(info)
  vim.notify(
    ("AzPim: open %s and enter code %s"):format(info.verification_uri, info.user_code),
    vim.log.levels.WARN,
    { title = "AzPim" }
  )
end

--- Called once a token has been obtained, so the UI can clear any prompt.
function M.on_authenticated() end

-- ---------------------------------------------------------------------------
-- token cache
-- ---------------------------------------------------------------------------

local session = {
  access_token = nil,
  expires_at = 0,
  refresh_token = nil,
  tenant = nil,
  loaded = false,
}

local function cache_path()
  return vim.fs.joinpath(vim.fn.stdpath("cache"), "azpim-graph-token.json")
end

local function load_cache()
  if session.loaded then
    return
  end
  session.loaded = true
  local fh = io.open(cache_path(), "r")
  if not fh then
    return
  end
  local raw = fh:read("*a")
  fh:close()
  local ok, data = pcall(vim.json.decode, raw)
  if ok and type(data) == "table" then
    session.access_token = data.access_token
    session.expires_at = tonumber(data.expires_at) or 0
    session.refresh_token = data.refresh_token
  end
end

--- Persist tokens at 0600 — this file can mint privileged Graph tokens.
local function save_cache()
  local fd = vim.uv.fs_open(cache_path(), "w", 384) -- 0600
  if not fd then
    return
  end
  vim.uv.fs_write(
    fd,
    vim.json.encode({
      access_token = session.access_token,
      expires_at = session.expires_at,
      refresh_token = session.refresh_token,
    })
  )
  vim.uv.fs_close(fd)
end

--- Forget the cached tokens; the next call signs in again.
function M.logout()
  session.access_token, session.refresh_token, session.expires_at = nil, nil, 0
  session.loaded = true
  os.remove(cache_path())
end

local function store(data)
  session.access_token = data.access_token
  session.expires_at = os.time() + (tonumber(data.expires_in) or 3600)
  if data.refresh_token then
    session.refresh_token = data.refresh_token
  end
  save_cache()
  vim.schedule(M.on_authenticated)
  return session.access_token, nil
end

local curl = http.curl
local encode_query = http.encode_query

-- ---------------------------------------------------------------------------
-- device code flow
-- ---------------------------------------------------------------------------

--- Sign in to the same tenant the CLI is pointed at, so the two agree.
local function tenant(cb)
  if session.tenant then
    return cb(session.tenant)
  end
  local ok = pcall(
    vim.system,
    { "az", "account", "show", "--query", "tenantId", "-o", "tsv" },
    { text = true },
    function(out)
      vim.schedule(function()
        local t = vim.trim(out.stdout or "")
        session.tenant = (out.code == 0 and t ~= "") and t or "organizations"
        cb(session.tenant)
      end)
    end
  )
  if not ok then
    session.tenant = "organizations"
    vim.schedule(function()
      cb(session.tenant)
    end)
  end
end

local function token_endpoint(t)
  return AUTHORITY .. "/" .. t .. "/oauth2/v2.0/token"
end

local function poll_device_code(t, device_code, interval, deadline, cb)
  vim.defer_fn(function()
    curl({
      "-X",
      "POST",
      token_endpoint(t),
      "--data-urlencode",
      "grant_type=urn:ietf:params:oauth:grant-type:device_code",
      "--data-urlencode",
      "client_id=" .. M.client_id,
      "--data-urlencode",
      "device_code=" .. device_code,
    }, function(data, err)
      if err then
        return cb(nil, err)
      end
      if data.access_token then
        return cb(store(data))
      end
      if data.error == "authorization_pending" or data.error == "slow_down" then
        if os.time() > deadline then
          return cb(nil, "device-code sign-in timed out")
        end
        local next_interval = interval + (data.error == "slow_down" and 5 or 0)
        return poll_device_code(t, device_code, next_interval, deadline, cb)
      end
      cb(nil, data.error_description or data.error or "device-code sign-in failed")
    end)
  end, interval * 1000)
end

local function device_flow(cb)
  tenant(function(t)
    curl({
      "-X",
      "POST",
      AUTHORITY .. "/" .. t .. "/oauth2/v2.0/devicecode",
      "--data-urlencode",
      "client_id=" .. M.client_id,
      "--data-urlencode",
      "scope=" .. table.concat(SCOPES, " "),
    }, function(data, err)
      if err then
        return cb(nil, err)
      end
      if not data.device_code then
        return cb(nil, data.error_description or data.error or "could not start device-code sign-in")
      end
      M.on_device_code({
        user_code = data.user_code,
        verification_uri = data.verification_uri or "https://login.microsoft.com/device",
        expires_in = data.expires_in,
      })
      poll_device_code(t, data.device_code, data.interval or 5, os.time() + (data.expires_in or 900), cb)
    end)
  end)
end

local function refresh_flow(cb)
  tenant(function(t)
    curl({
      "-X",
      "POST",
      token_endpoint(t),
      "--data-urlencode",
      "grant_type=refresh_token",
      "--data-urlencode",
      "client_id=" .. M.client_id,
      "--data-urlencode",
      "refresh_token=" .. session.refresh_token,
      "--data-urlencode",
      "scope=" .. table.concat(SCOPES, " "),
    }, function(data, err)
      if err then
        return cb(nil, err)
      end
      if data.access_token then
        return cb(store(data))
      end
      -- Expired or revoked refresh token: sign in again rather than fail.
      session.refresh_token = nil
      device_flow(cb)
    end)
  end)
end

-- A refresh triggers several Graph calls at once; without this queue each one
-- would start its own device-code flow.
local waiters = nil

---@param cb fun(token: string|nil, err: string|nil)
function M.token(cb)
  load_cache()
  if session.access_token and session.expires_at - 60 > os.time() then
    return cb(session.access_token, nil)
  end
  if waiters then
    return table.insert(waiters, cb)
  end
  waiters = { cb }
  local function finish(token, err)
    local pending = waiters
    waiters = nil
    for _, f in ipairs(pending) do
      f(token, err)
    end
  end
  if session.refresh_token then
    refresh_flow(finish)
  else
    device_flow(finish)
  end
end

--- True once a token is cached, i.e. no sign-in prompt is imminent.
function M.authenticated()
  load_cache()
  return session.refresh_token ~= nil or session.access_token ~= nil
end

-- ---------------------------------------------------------------------------
-- requests
-- ---------------------------------------------------------------------------

local function graph_error(data)
  local e = data.error
  if type(e) == "table" then
    return e.message or e.code or vim.inspect(e)
  end
  return data.error_description or tostring(e)
end

---@param cb fun(data: table|nil, err: string|nil)
function M.request(method, url, body, cb)
  M.token(function(token, err)
    if err then
      return cb(nil, err)
    end
    local args = {
      "-X",
      method:upper(),
      encode_query(url),
      "-H",
      "Authorization: Bearer " .. token,
    }
    local path
    if body then
      path = vim.fn.tempname() .. ".json"
      local fh, ferr = io.open(path, "w")
      if not fh then
        return cb(nil, "could not write request body: " .. tostring(ferr))
      end
      fh:write(vim.json.encode(body))
      fh:close()
      vim.list_extend(args, { "-H", "Content-Type: application/json", "--data-binary", "@" .. path })
    end
    curl(args, function(data, cerr, status)
      if path then
        os.remove(path)
      end
      if cerr then
        return cb(nil, cerr)
      end
      if data.error then
        return cb(nil, graph_error(data))
      end
      -- Graph normally sends `{"error":{…}}`, but a throttle/gateway failure can
      -- come back as a bare 429/503 with a body we'd otherwise read as success.
      local herr = http.http_error(data, status, "Graph")
      if herr then
        return cb(nil, herr)
      end
      cb(data, nil)
    end)
  end)
end

--- GET a URL, following `@odata.nextLink` until exhausted.
function M.get_all(url, cb, acc)
  acc = acc or {}
  M.request("get", url, nil, function(data, err)
    if err then
      return cb(nil, err)
    end
    vim.list_extend(acc, data.value or {})
    if data["@odata.nextLink"] then
      M.get_all(data["@odata.nextLink"], cb, acc)
    else
      cb(acc, nil)
    end
  end)
end

return M
